### PR TITLE
Pin all head objects into cache pool

### PR DESCRIPTION
**Pull request by wwformat**
When "cache_pin_head" is enabled, all the head objects will be promoted into cache pool, and the snap objects will be flushed into base pool.
In rbd case, all of head image data will keep in cache pool, the normal read/write ops will be hit and handle by cache pool directly, without extra cost of cache tier.
This mode is useful for store cache pool(replicated) and base pool(ec) into one disk.

### CODE REVIEW

**Comment by dragonylffly** [comment]
Does the head objects remain in the base pool? If so, they occupy space, and the inconsistence of head objects between base pool and cache pool seems not elegant. Besides, will the newly created snapshot objects  stay in the cache pool? should they be explicitly evicted

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -13967,6 +13984,22 @@ bool PrimaryLogPG::agent_choose_mode(bool restart, OpRequestRef op)
...
+  if (pool.info.cache_mode == pg_pool_t::CACHEMODE_WRITEBACK &&
+      pool.info.cache_pin_head) {
+    // do not flush omap objects if ec backing pool
+    const pg_pool_t *base_pool = get_osdmap()->get_pg_pool(pool.info.tier_of);
+    assert(base_pool);
+    uint64_t unflushable = base_pool->supports_omap() ?
+      0 : info.stats.stats.sum.num_objects_omap;
+
+    if (info.stats.stats.sum.num_whiteouts > 0 ||
+        info.stats.stats.sum.num_object_clones > unflushable) {
```
Comment (dragonylffly) [inline src/osd/PrimaryLogPG.cc]: why compare num_object_clones with num_objects_omap, why not num_object_clones > 0, what is the relationship with them

**Comment by wwformat** [comment]
> Does the head objects remain in the base pool? 

Yes, if the base pool has data already when the cache tier setup, but this is just for compatibility.
The head objects will never be flush to base pool if cache_pin_head is enabled.

>If so, they occupy space, and the inconsistence of head objects between base pool and cache pool seems not elegant. 

If the base pool is empty at the beginning, only snapshot objects will be flushed to base pool.

>Besides, will the newly created snapshot objects stay in the cache pool? should they be explicitly evicted

The new snapshot objects will stay in the cache pool in the beginning, and waiting for flush and evict.

File: src/osd/PrimaryLogPG.cc (hunk 1 above)
Comment (wwformat) [reply src/osd/PrimaryLogPG.cc]: the ec pool is not supported for omap

**Comment by dragonylffly** [comment]
>> If so, they occupy space, and the inconsistence of head objects between base pool and cache pool seems not elegant.
> 
> If the base pool is empty at the beginning, only snapshot objects will be flushed to base pool.

You mean add an empty base pool, and let the original pool as the cache pool? currently, although an nonempty cache pool is allowed to be added with --force-nonempty option, but if the original pool has snapshot, it is forbidden to do so. Otherwise, In the real world, the base pool is not empty at the beginning, and it is not erasure coded, so we can not enjoy the space benefit from this approach

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -13631,6 +13641,13 @@ bool PrimaryLogPG::agent_work(int start_max, int agent_flush_quota)
       osd->logger->inc(l_osd_agent_skip);
       continue;
     }
+    if (pool.info.cache_mode == pg_pool_t::CACHEMODE_WRITEBACK &&
+        pool.info.cache_pin_head &&
+        p->is_head() && !obc->obs.oi.is_whiteout()) {
+      dout(20) << __func__ << " skip (head) " << obc->obs.oi.soid << dendl;
+      osd->logger->inc(l_osd_agent_skip);
+      continue;
+    }
```
Comment (dragonylffly) [inline src/osd/PrimaryLogPG.cc]: can use FLAG_CACHE_PIN to avoid this?

**Comment by dragonylffly** [comment]
At current step, we can deal only with the new cluster steup. In that case, the base pool is empty and erasure coded. But we should write bypass cache pool for snapshot objects. Also, I think a new cache mode is better

File: src/osd/PrimaryLogPG.cc (hunk 2 above)
Comment (dragonylffly) [reply src/osd/PrimaryLogPG.cc]: can set this flag on the head objects?

**Comment by wwformat** [comment]
> I think we can use a brand new cache mode to explicitly seperate snapshots from heads in two pools. 

I like this, but current is simple.

> the current behavior is kind of strange, and the content of head objects in two pools are inconsistent, which, against the normal cache semantics. This essentially is not a cache. 

I think this is just a promote & flush & evict policy of writeback mode, for the huge cache pool cases which can be store all head objects.

> And to be practical, we should consider a way to depromote the snapshot objects into a new erasure coded pool, rather than promoting the head objects, since the original pool normally is replicated pool, keep snapshots in it will not gain space benefit. And after depromoting, delete the snapshot objects from the original pool. For the new snapshot objects after tiering, write bypass the original pool

If the can setup a writeback cache tier with cache_pin_head for cache pool(rep, nonempty) and base pool(ec, empty), the snapshot objects will be flush to ec pool and evict from cache pool by agents.

File: src/osd/PrimaryLogPG.cc (hunk 2 above)
Comment (wwformat) [reply src/osd/PrimaryLogPG.cc]: yes, but it is a rados api for get/set pin-flag, not for admin only.

**Pull request closed without merging**